File: cqrs-common/java/src-gen/main/java/org/fuin/dsl/cqrs/common/basics/MediaType.java
package org.fuin.dsl.cqrs.common.basics;

import jakarta.annotation.Generated;
import jakarta.validation.Constraint;
import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import jakarta.validation.Payload;
import java.io.Serial;
import java.io.Serializable;
import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.util.Objects;
import javax.annotation.concurrent.Immutable;
import org.fuin.objects4j.common.AsStringCapable;
import org.fuin.objects4j.common.ConstraintViolationException;
import org.fuin.objects4j.common.HasPublicStaticIsValidMethod;
import org.fuin.objects4j.common.HasPublicStaticValueOfMethod;
import org.fuin.objects4j.common.ValueObjectWithBaseType;

/**
 * A media type (formerly known as a Multipurpose Internet Mail Extensions or MIME type)     indicates the nature and format of a document, file, or assortment of bytes. MIME     types are defined and standardized in IETF's RFC 6838. The format is something like:     "type/subtype;parameter=value" (Example: "text/plain" or "application/json; encoding=UTF-8; version=1").
 */
@Immutable
@Generated("Generated class - Manual changes will be overwritten")
@HasPublicStaticIsValidMethod
@HasPublicStaticValueOfMethod
public final class MediaType implements ValueObjectWithBaseType<String>, Comparable<MediaType>, Serializable, AsStringCapable {

    @Serial
    private static final long serialVersionUID = 1000L;

    private static final int MAX_LENGTH = 100;

    @MediaTypeStr
    private String value;

    /**
     * Protected default constructor for deserialization.
     */
    @SuppressWarnings("NullAway.Init")
    protected MediaType() {
        super();
    }

    /**
     * Constructor with mandatory data.
     * 
     * @param value
     *            Value.
     */
    public MediaType(final String value) {
        super();
        requireArgValid("value", value);
        this.value = value;
    }

    @Override
    public String asBaseType() {
        return value;
    }

    @Override
    public String toString() {
        return value;
    }

    @Override
    public String asString() {
        return value;
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final MediaType other = (MediaType) obj;
        return Objects.equals(value, other.value);
    }

    @Override
    public int compareTo(final MediaType other) {
        return value.compareTo(other.value);
    }

    @Override
    public Class<String> getBaseType() {
        return String.class;
    }

    /**
     * Verifies that a given string can be converted into the type.
     * 
     * @param value
     *            Value to validate.
     * 
     * @return Returns {@literal true} if it's a valid type else {@literal false}.
     */
    public static boolean isValid(final String value) {
        if (value == null) {
            return true;
        }
        if (value.isEmpty()) {
            return false;
        }
        final String trimmed = value.trim();
        return trimmed.length() <= MAX_LENGTH;
    }

    /**
     * Verifies if the argument is valid and throws an exception if this is not
     * the case.
     * 
     * @param name
     *            Name of the value for a possible error message.
     * @param value
     *            Value to check.
     * 
     * @throws ConstraintViolationException
     *             The value was not valid.
     */
    public static void requireArgValid(final String name, final String value) throws ConstraintViolationException {

        if (!isValid(value)) {
            throw new ConstraintViolationException("The argument '" + name
                    + "' is not valid: '" + value + "'");
        }

    }

    /**
     * Ensures that the string can be converted into the type.
     */
    @Target({ ElementType.METHOD, ElementType.PARAMETER, ElementType.FIELD,
            ElementType.ANNOTATION_TYPE })
    @Retention(RetentionPolicy.RUNTIME)
    @Constraint(validatedBy = { Validator.class })
    @Documented
    public @interface MediaTypeStr {

        String message()

        default "{org.fuin.dsl.cqrs.common.basics.MediaType.message}";

        Class<?>[] groups() default {};

        Class<? extends Payload>[] payload() default {};

    }

    /**
     * Validates if a string is compliant with the type.
     */
    public static final class Validator implements
            ConstraintValidator<MediaTypeStr, String> {

        @Override
        public void initialize(
                final MediaTypeStr annotation) {
            // Not used
        }

        @Override
        public boolean isValid(final String value,
                final ConstraintValidatorContext context) {
            return MediaType.isValid(value);
        }

    }


}

File: cqrs-common/dart/lib/src-gen/common/basics/media_type.dart
import 'package:cqrs_common/src/descriptor/constraint.dart';

/// A media type (formerly known as a Multipurpose Internet Mail Extensions or MIME type) indicates
/// the nature and format of a document, file, or assortment of bytes. MIME types are defined and
/// standardized in IETF's RFC 6838. The format is something like: "type/subtype;parameter=value"
/// (Example: "text/plain" or "application/json; encoding=UTF-8; version=1").
class MediaType {
  /// Constructor with mandatory data.
  const MediaType(this.value);

  /// Reads the value off the wire, where a wrapper of one value travels as that value.
  ///
  /// A single-value object is not an object on the wire - the server writes the value itself -
  /// so there is nothing here for a `fromJson` to read a field out of. This is the same door an
  /// id and an enum offer, for the same reason.
  factory MediaType.fromWire(Object value) => MediaType(value as String);

  /// The invariants the model declares on this type, as far as this target can express them.
  ///
  /// Always present, even when empty. Whether a constraint has a Dart equivalent is this
  /// target's business, not a caller's: a descriptor naming this type's constraints has to
  /// compile whatever the model happens to declare, and a member that appears and disappears
  /// makes one factory's output depend on another's.
  static const List<Constraint> constraints = <Constraint>[Length(3, 127)];

  final String value;

  /// Whether [value] satisfies the model's invariants.
  static bool isValid(Object? value) => validate(value, 'MediaType') == null;

  /// What is wrong with [value], or `null` when it satisfies the model's invariants.
  ///
  /// [label] is what to call the value in the message - a field's label from the model, so the
  /// wording a user reads is the wording the model states.
  static String? validate(Object? value, String label) {
    for (final constraint in constraints) {
      final message = constraint.validate(value, label);
      if (message != null) {
        return message;
      }
    }
    return null;
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is MediaType && other.value == value;

  @override
  int get hashCode => value.hashCode;

  @override
  String toString() => value.toString();
}
